adapters/repos/db/lsmkv/bucket_reader_roaring_set_range.go

```diff
@@ -0,0 +1,363 @@
+		return resultBM, err
+	}
+
+	// all values are >= 0
+	if value == 0 {
+		return resultBM, nil
+	}
+
+	return r.mergeGreaterThanEqual(ctx, resultBM, cursor, value)
+}
+
+func (r *BucketReaderRoaringSetRange) greaterThan(ctx context.Context, value uint64) (*sroar.Bitmap, error) {
+	// no value is > max uint64
+	if value == math.MaxUint64 {
+		return sroar.NewBitmap(), nil
+	}
+
+	resultBM, cursor, ok, err := r.nonNullBMWithCursor(ctx)
+	if !ok {
+		return resultBM, err
+	}
+
+	return r.mergeGreaterThanEqual(ctx, resultBM, cursor, value+1)
+}
+
+func (r *BucketReaderRoaringSetRange) lessThanEqual(ctx context.Context, value uint64) (*sroar.Bitmap, error) {
+	resultBM, cursor, ok, err := r.nonNullBMWithCursor(ctx)
+	if !ok {
+		return resultBM, err
+	}
+
+	// all values are <= max uint64
+	if value == math.MaxUint64 {
+		return resultBM, nil
+	}
+
+	greaterThanEqualBM, err := r.mergeGreaterThanEqual(ctx, resultBM.Clone(), cursor, value+1)
+	if err != nil {
+		return nil, err
+	}
+	resultBM.AndNot(greaterThanEqualBM)
+	return resultBM, nil
+}
+
+func (r *BucketReaderRoaringSetRange) lessThan(ctx context.Context, value uint64) (*sroar.Bitmap, error) {
+	// no value is < 0
+	if value == 0 {
+		return sroar.NewBitmap(), nil
+	}
+
+	resultBM, cursor, ok, err := r.nonNullBMWithCursor(ctx)
+	if !ok {
+		return resultBM, err
+	}
+
+	greaterThanEqualBM, err := r.mergeGreaterThanEqual(ctx, resultBM.Clone(), cursor, value)
+	if err != nil {
+		return nil, err
+	}
+	resultBM.AndNot(greaterThanEqualBM)
+	return resultBM, nil
+}
+
+func (r *BucketReaderRoaringSetRange) equal(ctx context.Context, value uint64) (*sroar.Bitmap, error) {
+	if value == 0 {
+		return r.lessThanEqual(ctx, value)
+	}
+	if value == math.MaxUint64 {
+		return r.greaterThanEqual(ctx, value)
+	}
+
+	resultBM, cursor, ok, err := r.nonNullBMWithCursor(ctx)
+	if !ok {
+		return resultBM, err
+	}
+
+	return r.mergeEqual(ctx, resultBM, cursor, value)
+}
+
+func (r *BucketReaderRoaringSetRange) notEqual(ctx context.Context, value uint64) (*sroar.Bitmap, error) {
+	if value == 0 {
+		return r.greaterThan(ctx, value)
+	}
+	if value == math.MaxUint64 {
+		return r.lessThan(ctx, value)
+	}
+
+	resultBM, cursor, ok, err := r.nonNullBMWithCursor(ctx)
+	if !ok {
+		return resultBM, err
+	}
+
+	equalBM, err := r.mergeEqual(ctx, resultBM.Clone(), cursor, value)
+	if err != nil {
+		return nil, err
+	}
+	resultBM.AndNot(equalBM)
+	return resultBM, nil
+}
+
+func (r *BucketReaderRoaringSetRange) nonNullBMWithCursor(ctx context.Context) (*sroar.Bitmap, *noGapsCursor, bool, error) {
+	cursor := &noGapsCursor{cursor: r.cursorFn()}
+	_, nonNullBM, _ := cursor.first()
+
+	// if non-null bm is nil or empty, no values are present
+	if nonNullBM == nil || nonNullBM.IsEmpty() {
+		cursor.close()
+		return sroar.NewBitmap(), nil, false, nil
+	}
+
+	if ctx.Err() != nil {
+		cursor.close()
+		return nil, nil, false, ctx.Err()
+	}
+
+	return nonNullBM, cursor, true, nil
+	// return nonNullBM.Clone(), cursor, true, nil
```

Comment:
Do we not need to clone here anymore?

Reply:
nope. cursor should take care of cloning or bitmaps that are going to be mutated in the process